pretyping/evarsolve.ml

```diff
@@ -73,25 +72,33 @@ let refresh_universes ?(status=univ_rigid) ?(onlyalg=false) ?(refreshset=false)
        refresh_sort status ~direction s
       | _ -> t
       end
-    | Prod (na,u,v) -> 
-      mkProd (na, u, refresh ~onlyalg status ~direction v)
+    | Prod (na,u,v) ->
+       let v' = refresh ~onlyalg status ~direction v in
+       if v' == v then t else mkProd (na, u, v')
     | _ -> t
   (** Refresh the types of evars under template polymorphic references *)
   and refresh_term_evars onevars top t =
     match EConstr.kind !evdref t with
     | App (f, args) when is_template_polymorphic env !evdref f ->
       let pos = get_polymorphic_positions !evdref f in
-	refresh_polymorphic_positions args pos
+        refresh_polymorphic_positions args pos; t
     | App (f, args) when top && isEvar !evdref f -> 
-      refresh_term_evars true false f; 
-      Array.iter (refresh_term_evars onevars false) args
+       let f' = refresh_term_evars true false f in
+       let args' = Array.map (refresh_term_evars onevars false) args in
+       if f' == f && args' == args then t
+       else mkApp (f', args')
     | Evar (ev, a) when onevars ->
       let evi = Evd.find !evdref ev in
-      let ty' = refresh ~onlyalg univ_flexible ~direction:true (EConstr.of_constr evi.evar_concl) in
-	if !modified then 
-	  evdref := Evd.add !evdref ev {evi with evar_concl = EConstr.Unsafe.to_constr ty'}
-	else ()
-    | _ -> EConstr.iter !evdref (refresh_term_evars onevars false) t
+      let ty = EConstr.of_constr evi.evar_concl in
+      let ty' = refresh ~onlyalg univ_flexible ~direction:true ty in
+      if ty == ty' then t
+      else (evdref := Evd.add !evdref ev {evi with evar_concl = EConstr.to_constr !evdref ty'}; t)
```

Comment:
`Evd.add` has a comment:
```ocaml
(** [add sigma ev info] adds [ev] with evar info [info] in sigma.
    Precondition: ev must not preexist in [sigma]. *)
```
Is it outdated or what?

Reply:
That's a good practice comment, but a few people break this invariant, *cough* ssr *cough*. We should provide a function for evar updates.

Reply:
@gares asks if this doesn't break the idea behind the monotonic interface, in the sense that you have `t` and `t'` in `sigma`, you refresh `t`, and get `t1` and `sigma1'. Now if you typecheck `t'` in `sigma1`, you crash.

What do you mean by an evar update function, in the context of such monotonicity invariants?

Reply:
Refreshing ensures that any solution of the initial evars is typable with the refreshed type if it was typable with the non-refreshed one. Replaced this with Evd.downcast which does precisely this.